### PR TITLE
Extract task queue

### PR DESCRIPTION
**Pull request by obrok**
Factors out a `TaskQueue` from `Member` so that `Member` itself doesn't need to deal with scheduling, just use the other component. The sum total of functionality is about the same, it's just divided more cleanly.

### CODE REVIEW

**Comment by mike1729** [comment]
Using `mem::take` here to cheat the borrow checker is very error prone, docs are not enough to guard against a potential mistake.
Slightly better way would be to have `task_queue` as an option in `Member` but this would make the rest of the code awful.
The split would be the proper way to make the compiler happy, but I'm not sure if it's worth it, since the declaration of `work-off` method shows how tangled the `task_queue` and the rest of the `Member` are with each other. On the other hand, it would be good to have the `Member` better organized. I'm in favor of split instead of abandoning if you know to do it in a sensible way, i.e. the split represents different functionalities not just `task_queue` and `rest` :)

**Comment by obrok** [comment]
Well, it's pretty clear to me that `task_queue` is a separate functionality and I don't agree it's entangled at all. I think it's just a rust quirk that makes it difficult to split here. Actually another version that I think could hide a reasonable amount of the functionality is to expose a `due_task -> Option<ScheduledTask>` and a `reschedule(ScheduledTask, delay)` functions. This leaks the `ScheduledTask` struct outside the `task_queue`, but avoids the problem with the borrow.

**Comment by timorl** [comment]
I'm all for splitting `member` into smaller pieces, but I'm pretty sure `work_off` is unnecessarily complex. Since `counter` only matters for a few tasks (and doesn't matter internally for `TaskQueue`) you could keep it inside the relevant tasks, and then all you could have `due_task` return `Option<Task>`, avoiding the leakage. After a task is processed it could return another task to be scheduled with a delay (either the same task, or one with an increased counter).

**Comment by obrok** [comment]
@timorl, yeah, that's a good point, I think that's the simplest approach.

**Comment by mike1729** [comment]
Yeah, that's a good approach :)

sidenote: this "rust quirk" is called borrow checker and people are generally fond of it :P

**Comment by obrok** [comment]
> sidenote: this "rust quirk" is called borrow checker and people are generally fond of it :P

I realize that, but - at the risk of being too philosophical here - the borrow checker could (for some definition of "could" :smile:) be smarter and detect when only part of a datastructure needs to be borrowed like in this case, instead of being manually convinced that that is indeed what's happening. I more so meant this fact that the borrow checker only sees the whole datastructure when I said "a quirk".

**Comment by obrok** [comment]
@mike1729 @timorl here's how it looks with the counter logic moved back into `member`. I also changed it to `rescheduled: bool`, since that's what it effectively tracks now, it never cares how many time something happened, even for printing.

**Comment by timorl** [comment]
It actually does, but it has its limitations – in particular if you use a method on a struct, that method could in principle touch arbitrary parts of the struct. Propagating such information across method lines would almost surely result in terrible compiler performance.

**Comment by mike1729** [comment]
LGTM :) unwip?

**Comment by obrok** [comment]
I'm on vacation starting Monday, so I'll leave this in @DamianStraszak's capable hands. I merged main into this, so it should be pretty clean to apply. Note that this PR currently targets a feature branch, so should be merged after that branch.

**Comment by obrok** [comment]
@mike1729 @DamianStraszak I resolved conflicts with main, please take another look.

**Review by mike1729 — APPROVED** [review]
very halal!

**Review by DamianStraszak — APPROVED** [review]
Nice and clean. Please more :)

File: consensus/src/task_queue.rs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    /// tasks.
+    pub fn pop_due_task(&mut self) -> Option<T> {
+        if let Some(task) = self.queue.peek() {
+            if task.scheduled_time <= time::Instant::now() {
+                let task = self.queue.pop().expect("The element was peeked");
+                return Some(task.task);
+            }
+        }
+
+        None
```
[the remark this answers is not in the document]
Comment (obrok) [reply consensus/src/task_queue.rs]: Applied

**Pull request merged**